testsuite/testcases/src/compatibility_test.rs

```diff
@@ -33,85 +25,17 @@ impl Test for SimpleValidatorUpgrade {
     }
 }
 
-async fn stat_gather_task(
-    emitter: TxnEmitter,
-    emit_job_request: EmitJobRequest,
-    source_account: Arc<LocalAccount>,
-    upgrade_traffic_chunk_duration: Duration,
-    done: Arc<AtomicBool>,
-) -> Result<Option<TxnStats>> {
-    let mut upgrade_stats = vec![];
-    while !done.load(Ordering::Relaxed) {
-        info!("stat_gather_task some traffic...");
-        let upgrading_stats = emitter
-            .clone()
-            .emit_txn_for(
-                source_account.clone(),
-                emit_job_request.clone(),
-                upgrade_traffic_chunk_duration,
-            )
-            .await?;
-        info!("stat_gather_task some stats: {}", &upgrading_stats);
-        upgrade_stats.push(upgrading_stats);
-    }
-    let statsum = upgrade_stats.into_iter().reduce(|a, b| &a + &b);
-    Ok(statsum)
-}
-
-fn upgrade_and_gather_stats(
+fn upgrade(
     ctxa: NetworkContextSynchronizer,
     // upgrade args
     validators_to_update: &[PeerId],
     version: &Version,
     wait_until_healthy: bool,
     delay: Duration,
     max_wait: Duration,
-    // traffic args
-    nodes: &[PeerId],
-) -> Result<Option<TxnStats>> {
-    let upgrade_done = Arc::new(AtomicBool::new(false));
-    let emitter_ctx = ctxa.clone();
-    let mut stats_result: Result<Option<TxnStats>> = Ok(None);
+) -> Result<()> {
     let mut upgrade_result: Result<()> = Ok(());
     tokio_scoped::scope(|scopev| {
```

Comment:
Do you still need to spawn a separate thread here?

Reply:
Yes, this is still needed - there is a deadlock situation aparently if we are not spawning in the same context.